src/NuGet.Core/NuGet.Common/PathUtil/FileUtility.cs

```diff
@@ -12,7 +13,118 @@ public static class FileUtility
+        /// </summary>
+        public static string GetTempFilePath(string directory)
+        {
+            var fileName = $"{Guid.NewGuid()}.tmp".ToLowerInvariant();
+
+            return Path.GetFullPath(Path.Combine(directory, fileName));
+        }
+
+        /// <summary>
+        /// Lock around the output path.
+        /// Delete the existing file with retries.
+        /// </summary>
+        public static async Task DeleteWithLock(string filePath)
+        {
+            if (filePath == null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            await ConcurrencyUtilities.ExecuteWithFileLockedAsync(filePath,
+                lockedToken =>
+                {
+                    Delete(filePath);
+
+                    return Task.FromResult(0);
+                },
+                // Do not allow this to be cancelled
+                CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Lock around the output path.
+        /// Delete the existing file with retries.
+        /// Move a file with retries.
+        /// </summary>
+        public static async Task ReplaceWithLock(Action<string> writeSourceFile, string destFilePath)
+        {
+            if (writeSourceFile == null)
+            {
+                throw new ArgumentNullException(nameof(writeSourceFile));
+            }
+
+            if (destFilePath == null)
+            {
+                throw new ArgumentNullException(nameof(destFilePath));
+            }
+
+            await ConcurrencyUtilities.ExecuteWithFileLockedAsync(destFilePath,
+                lockedToken =>
+                {
+                    Replace(writeSourceFile, destFilePath);
+
+                    return Task.FromResult(0);
+                },
+                // Do not allow this to be cancelled
+                CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Delete the existing file with retries.
+        /// Move a file with retries.
+        /// </summary>
+        public static void Replace(Action<string> writeSourceFile, string destFilePath)
+        {
+            if (writeSourceFile == null)
+            {
+                throw new ArgumentNullException(nameof(writeSourceFile));
+            }
+
+            if (destFilePath == null)
+            {
+                throw new ArgumentNullException(nameof(destFilePath));
+            }
+
+            var tempPath = GetTempFilePath(Path.GetDirectoryName(destFilePath));
+
+            try
+            {
+                // Write to temp path
+                writeSourceFile(tempPath);
+
+                // Delete the previous file and move the temporary file
+                // This will throw if there is a failure
+                Replace(tempPath, destFilePath);
+            }
+            catch
+            {
+                // Clean up the temporary file
+                Delete(tempPath);
```

Comment:
@emgarten This will throw the wrong exception if Delete fails. We should catch around Delete (since it throws at the end) and in that case throw an aggregate exception with both exceptions or something along these lines, otherwise we are going to end up with a weird log